File: data/compilers_comparison/src/examples/first_example/src/func4.c
#include "first_example.h" 
array_t* func4(array_t_param* vars, const unsigned long PATH0, int loopsFactor) {
   size_t pCounter = vars->size;
   array_t* array110;
   if (pCounter > 0) {
      array110 = vars->data[--pCounter];
      array110->refC++;
      DEBUG_COPY(array110->id);
   } else {
      array110 = (array_t*)malloc(sizeof(array_t));
      array110->size = 209;
      array110->refC = 1;
      array110->id = 110;
      array110->data = (unsigned int*)malloc(array110->size*sizeof(unsigned int));
      memset(array110->data, 0, array110->size*sizeof(unsigned int));
      DEBUG_NEW(array110->id);
   }
   unsigned int loop33 = 0;
   unsigned int loopLimit33 = (rand()%loopsFactor)/3 + 1;
   for(; loop33 < loopLimit33; loop33++) {
      if(PATH0 & 1) {
         array_t* array111;
         if (pCounter > 0) {
            array111 = vars->data[--pCounter];
            array111->refC++;
            DEBUG_COPY(array111->id);
         } else {
            array111 = (array_t*)malloc(sizeof(array_t));
            array111->size = 581;
            array111->refC = 1;
            array111->id = 111;
            array111->data = (unsigned int*)malloc(array111->size*sizeof(unsigned int));
            memset(array111->data, 0, array111->size*sizeof(unsigned int));
            DEBUG_NEW(array111->id);
         }
         unsigned int loop34 = 0;
         unsigned int loopLimit34 = (rand()%loopsFactor)/4 + 1;
         for(; loop34 < loopLimit34; loop34++) {
            for (int i = 0; i < array111->size; i++) {
               array111->data[i]--;
            }
            for (int i = 0; i < array110->size; i++) {
               if (array110->data[i] == 32) { 
                  return array110;
               }
            }
         }
         array_t_param params0;
         params0.size = 2;
         params0.data = (array_t**)malloc(params0.size*sizeof(array_t*));
         params0.data[0] = array110;
         params0.data[1] = array111;
         array_t* array112 = func24(&params0, rng(), loopsFactor);
         DEBUG_RETURN(array112->id);
         free(params0.data);
         array112->refC--;
         if(array112->refC == 0) {
            free(array112->data);
            free(array112);
            DEBUG_FREE(array112->id);
         }
         array111->refC--;
         if(array111->refC == 0) {
            free(array111->data);
            free(array111);
            DEBUG_FREE(array111->id);
         }
      }
      else {
         array_t_param params0;
         params0.size = 1;
         params0.data = (array_t**)malloc(params0.size*sizeof(array_t*));
         params0.data[0] = array110;
         array_t* array113 = func16(&params0, rng(), loopsFactor);
         DEBUG_RETURN(array113->id);
         free(params0.data);
         array113->refC--;
         if(array113->refC == 0) {
            free(array113->data);
            free(array113);
            DEBUG_FREE(array113->id);
         }
      }
      array_t* array114;
      if (pCounter > 0) {
         array114 = vars->data[--pCounter];
         array114->refC++;
         DEBUG_COPY(array114->id);
      } else {
         array114 = (array_t*)malloc(sizeof(array_t));
         array114->size = 769;
         array114->refC = 1;
         array114->id = 114;
         array114->data = (unsigned int*)malloc(array114->size*sizeof(unsigned int));
         memset(array114->data, 0, array114->size*sizeof(unsigned int));
         DEBUG_NEW(array114->id);
      }
      array_t_param params0;
      params0.size = 2;
      params0.data = (array_t**)malloc(params0.size*sizeof(array_t*));
      params0.data[0] = array110;
      params0.data[1] = array114;
      array_t* array115 = func17(&params0, loopsFactor);
      DEBUG_RETURN(array115->id);
      free(params0.data);
      if(PATH0 & 2) {
         array_t* array116;
         if (pCounter > 0) {
            array116 = vars->data[--pCounter];
            array116->refC++;
            DEBUG_COPY(array116->id);
         } else {
            array116 = (array_t*)malloc(sizeof(array_t));
            array116->size = 850;
            array116->refC = 1;
            array116->id = 116;
            array116->data = (unsigned int*)malloc(array116->size*sizeof(unsigned int));
            memset(array116->data, 0, array116->size*sizeof(unsigned int));
            DEBUG_NEW(array116->id);
         }
         unsigned int loop35 = 0;
         unsigned int loopLimit35 = (rand()%loopsFactor)/4 + 1;
         for(; loop35 < loopLimit35; loop35++) {
            for (int i = 0; i < array116->size; i++) {
               array116->data[i]--;
            }
            for (int i = 0; i < array110->size; i++) {
               if (array110->data[i] == 42) { 
                  return array110;
               }
            }
         }
         array_t_param params1;
         params1.size = 4;
         params1.data = (array_t**)malloc(params1.size*sizeof(array_t*));
         params1.data[0] = array110;
         params1.data[1] = array114;
         params1.data[2] = array115;
         params1.data[3] = array116;
         array_t* array117 = func28(&params1, rng(), loopsFactor);
         DEBUG_RETURN(array117->id);
         free(params1.data);
         array117->refC--;
         if(array117->refC == 0) {
            free(array117->data);
            free(array117);
            DEBUG_FREE(array117->id);
         }
         array116->refC--;
         if(array116->refC == 0) {
            free(array116->data);
            free(array116);
            DEBUG_FREE(array116->id);
         }
      }
      else {
         array_t_param params1;
         params1.size = 3;
         params1.data = (array_t**)malloc(params1.size*sizeof(array_t*));
         params1.data[0] = array110;
         params1.data[1] = array114;
         params1.data[2] = array115;
         array_t* array118 = func22(&params1, rng(), loopsFactor);
         DEBUG_RETURN(array118->id);
         free(params1.data);
         array118->refC--;
         if(array118->refC == 0) {
            free(array118->data);
            free(array118);
            DEBUG_FREE(array118->id);
         }
      }
      for (int i = 0; i < array115->size; i++) {
         if (array115->data[i] == 67) { 
            return array115;
         }
      }
      array115->refC--;
      if(array115->refC == 0) {
         free(array115->data);
         free(array115);
         DEBUG_FREE(array115->id);
      }
      array114->refC--;
      if(array114->refC == 0) {
         free(array114->data);
         free(array114);
         DEBUG_FREE(array114->id);
      }
   }
   for (int i = 0; i < array110->size; i++) {
      array110->data[i]--;
   }
   for (int i = 0; i < array110->size; i++) {
      if (array110->data[i] == 61) { 
         return array110;
      }
   }
   return array110;
}

File: data/multilang_comparison/program_zig/src/func4.zig
const lib = @import("lib.zig");
const std = lib.std;
const allocator = std.heap.smp_allocator;
const Array = lib.Array;
const ArrayParam = lib.ArrayParam;
const get_path = lib.get_path;

const func0 = @import("func0.zig").func0;
const func10 = @import("func10.zig").func10;
const func11 = @import("func11.zig").func11;
const func28 = @import("func28.zig").func28;
const func29 = @import("func29.zig").func29;
const func40 = @import("func40.zig").func40;
const func41 = @import("func41.zig").func41;
const func30 = @import("func30.zig").func30;
const func42 = @import("func42.zig").func42;
const func43 = @import("func43.zig").func43;
const func44 = @import("func44.zig").func44;
const func45 = @import("func45.zig").func45;
const func34 = @import("func34.zig").func34;
const func31 = @import("func31.zig").func31;
const func35 = @import("func35.zig").func35;
const func22 = @import("func22.zig").func22;
const func23 = @import("func23.zig").func23;
const func36 = @import("func36.zig").func36;
const func37 = @import("func37.zig").func37;
const func38 = @import("func38.zig").func38;
const func39 = @import("func39.zig").func39;
const func32 = @import("func32.zig").func32;
const func12 = @import("func12.zig").func12;
const func24 = @import("func24.zig").func24;
const func25 = @import("func25.zig").func25;
const func26 = @import("func26.zig").func26;
const func27 = @import("func27.zig").func27;
const func33 = @import("func33.zig").func33;
const func16 = @import("func16.zig").func16;
const func13 = @import("func13.zig").func13;
const func17 = @import("func17.zig").func17;
const func5 = @import("func5.zig").func5;
const func18 = @import("func18.zig").func18;
const func19 = @import("func19.zig").func19;
const func20 = @import("func20.zig").func20;
const func21 = @import("func21.zig").func21;
const func14 = @import("func14.zig").func14;
const func6 = @import("func6.zig").func6;
const func7 = @import("func7.zig").func7;
const func8 = @import("func8.zig").func8;
const func9 = @import("func9.zig").func9;
const func15 = @import("func15.zig").func15;
const func2 = @import("func2.zig").func2;
const func1 = @import("func1.zig").func1;
const func3 = @import("func3.zig").func3;

pub fn func4(vars: *ArrayParam, _path0: u64) !*Array {
   var pCounter: usize = vars.size;
   _ = &pCounter;
   const path0 = _path0;
   _ = &path0;
   var array110: *Array = undefined;
   if (pCounter > 0) {
       pCounter -= 1;
       array110 = vars.data[pCounter];
       array110.refC += 1;
   } else {
       array110 = try allocator.create(Array);
       array110.*.size = 209;
       array110.*.refC = 1;
       array110.*.id = 110;
       array110.*.data = try allocator.alloc(u32, array110.*.size);
   }
   var loop33: usize = 0;
   const loopLimit33: usize = (50)/3 + 1;
   while (loop33 < loopLimit33) {
      if ((path0 & 1) != 0) {
         var array111: *Array = undefined;
         if (pCounter > 0) {
             pCounter -= 1;
             array111 = vars.data[pCounter];
             array111.refC += 1;
         } else {
             array111 = try allocator.create(Array);
             array111.*.size = 581;
             array111.*.refC = 1;
             array111.*.id = 111;
             array111.*.data = try allocator.alloc(u32, array111.*.size);
         }
         var loop34: usize = 0;
         const loopLimit34: usize = (50)/4 + 1;
         while (loop34 < loopLimit34) {
            for (0..array111.size) |i| {
                array111.data[i] -= 1;
            }
            for (0..array110.size) |i| {
                if (array110.data[i] == 32) {
                    return array110;
                }
            }
            loop34 += 1;
         }
         const params0_data = try allocator.alloc(*Array, 2);
         params0_data[0] = array110;
         params0_data[1] = array111;
         var params0 = ArrayParam{
             .data = params0_data,
             .size = 2,
         };
         const array112 = try func24(&params0, try get_path());
         defer allocator.free(params0.data);
         array112.refC -= 1;
         if (array112.refC == 0) {
             allocator.free(array112.data);
             allocator.destroy(array112);
         }
         array111.refC -= 1;
         if (array111.refC == 0) {
             allocator.free(array111.data);
             allocator.destroy(array111);
         }
      
      } else {
         const params0_data = try allocator.alloc(*Array, 1);
         params0_data[0] = array110;
         var params0 = ArrayParam{
             .data = params0_data,
             .size = 1,
         };
         const array113 = try func16(&params0, try get_path());
         defer allocator.free(params0.data);
         array113.refC -= 1;
         if (array113.refC == 0) {
             allocator.free(array113.data);
             allocator.destroy(array113);
         }
      }
      var array114: *Array = undefined;
      if (pCounter > 0) {
          pCounter -= 1;
          array114 = vars.data[pCounter];
          array114.refC += 1;
      } else {
          array114 = try allocator.create(Array);
          array114.*.size = 769;
          array114.*.refC = 1;
          array114.*.id = 114;
          array114.*.data = try allocator.alloc(u32, array114.*.size);
      }
      const params0_data = try allocator.alloc(*Array, 2);
      params0_data[0] = array110;
      params0_data[1] = array114;
      var params0 = ArrayParam{
          .data = params0_data,
          .size = 2,
      };
      const array115 = try func17(&params0);
      defer allocator.free(params0.data);
      if ((path0 & 2) != 0) {
         var array116: *Array = undefined;
         if (pCounter > 0) {
             pCounter -= 1;
             array116 = vars.data[pCounter];
             array116.refC += 1;
         } else {
             array116 = try allocator.create(Array);
             array116.*.size = 850;
             array116.*.refC = 1;
             array116.*.id = 116;
             array116.*.data = try allocator.alloc(u32, array116.*.size);
         }
         var loop35: usize = 0;
         const loopLimit35: usize = (50)/4 + 1;
         while (loop35 < loopLimit35) {
            for (0..array116.size) |i| {
                array116.data[i] -= 1;
            }
            for (0..array110.size) |i| {
                if (array110.data[i] == 42) {
                    return array110;
                }
            }
            loop35 += 1;
         }
         const params1_data = try allocator.alloc(*Array, 4);
         params1_data[0] = array110;
         params1_data[1] = array114;
         params1_data[2] = array115;
         params1_data[3] = array116;
         var params1 = ArrayParam{
             .data = params1_data,
             .size = 4,
         };
         const array117 = try func28(&params1, try get_path());
         defer allocator.free(params1.data);
         array117.refC -= 1;
         if (array117.refC == 0) {
             allocator.free(array117.data);
             allocator.destroy(array117);
         }
         array116.refC -= 1;
         if (array116.refC == 0) {
             allocator.free(array116.data);
             allocator.destroy(array116);
         }
      
      } else {
         const params1_data = try allocator.alloc(*Array, 3);
         params1_data[0] = array110;
         params1_data[1] = array114;
         params1_data[2] = array115;
         var params1 = ArrayParam{
             .data = params1_data,
             .size = 3,
         };
         const array118 = try func22(&params1, try get_path());
         defer allocator.free(params1.data);
         array118.refC -= 1;
         if (array118.refC == 0) {
             allocator.free(array118.data);
             allocator.destroy(array118);
         }
      }
      for (0..array115.size) |i| {
          if (array115.data[i] == 67) {
              return array115;
          }
      }
      array115.refC -= 1;
      if (array115.refC == 0) {
          allocator.free(array115.data);
          allocator.destroy(array115);
      }
      array114.refC -= 1;
      if (array114.refC == 0) {
          allocator.free(array114.data);
          allocator.destroy(array114);
      }
      loop33 += 1;
   }
   for (0..array110.size) |i| {
       array110.data[i] -= 1;
   }
   for (0..array110.size) |i| {
       if (array110.data[i] == 61) {
           return array110;
       }
   }
   return array110;
}

